### DISCUSSION TITLE
How to enable cython line tracing?

### DISCUSSION THREAD

**Discussion by ghost** [mpi4py/mpi4py#122] [Q&A] [answered]
I'd like to have some coverage reports so I tried the `Cython.Coverage` plugin for coverage.py but I can't use it. It needs line tracing enabled.
https://cython.readthedocs.io/en/latest/src/tutorial/profiling_tutorial.html#enabling-coverage-analysis
https://cython.readthedocs.io/en/latest/src/tutorial/profiling_tutorial.html#enabling-line-tracing

- In file ` src/mpi4py/MPI.pyx` I add `# cython: linetrace=True`
- I think mpi4py uses a customized `cythonize.py` so `# distutils: define_macros=CYTHON_TRACE_NOGIL=1` has no effect.
- Cython docs alternatively suggest to modify `setup.py`. I tried `ext.define_macros += [('CYTHON_TRACE_NOGIL', '1')]` inside `configure_mpi` function.

I get `Couldn't parse 'src/mpi4py/MPI.pyx' as Python source: 'invalid syntax' at line 13`. Line 13 is just `cimport cython`

**Comment by dalcinl** [comment, marked as answer]
Last time I tried (years ago) I could not make it work. I do not remember exactly what the problems were, but it may have been related to the (ab)use of "include" in mpi4py sources. 

Do not use the `# distutils: ... `comments, they will be ignored. Defining a preprocessor macro in `setup.py` should work, you could even use `export CFLAGS=-DCYTHON_TRACE_NOGIL=1`. I tried (together with the adition of `#cython: linetrace=True` at the beginning of `MPI.pyx`), and the C code seems to compile just fine. Please double check, I'm sure you have some trivial mistake around.

**Reply by ghost** [reply]
Solution: Use `CFLAGS=-DCYTHON_TRACE=1` instead.
1. Download latest mpi4py and add `#cython: linetrace=True` at the beginning of `src/mpi4py/MPI.pyx`
2. Compile with `CFLAGS=-DCYTHON_TRACE=1 make`

Don't use `CFLAGS=-DCYTHON_TRACE_NOGIL=1` as it segfaults on a simple `mpiexec -n 2 python  demo/helloworld.py`.

Sorry I wasn't precise enough. The parsing error above is not relevant because there was a segfault before.

The parsing error is a known bug in Cython.Coverage plugin
https://github.com/cython/cython/issues/1508
https://github.com/cython/cython/issues/3636
